core/src/main/java/tc/oc/pgm/spawner/Spawner.java

```diff
@@ -0,0 +1,134 @@
+package tc.oc.pgm.spawner;
+
+import org.bukkit.Effect;
+import org.bukkit.entity.Entity;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.EntityDeathEvent;
+import org.bukkit.event.entity.ItemDespawnEvent;
+import org.bukkit.event.player.PlayerPickupItemEvent;
+import org.bukkit.util.Vector;
+import tc.oc.pgm.api.PGM;
+import tc.oc.pgm.api.event.CoarsePlayerMoveEvent;
+import tc.oc.pgm.api.feature.Feature;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.Tickable;
+import tc.oc.pgm.api.match.event.MatchFinishEvent;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.api.time.Tick;
+import tc.oc.pgm.util.TimeUtils;
+import tc.oc.pgm.util.bukkit.OnlinePlayerMapAdapter;
+
+public class Spawner implements Feature<SpawnerDefinition>, Listener, Tickable {
+
+  public static final String METADATA_KEY = "spawner";
+
+  private final Match match;
+  private final SpawnerDefinition definition;
+  private final OnlinePlayerMapAdapter<MatchPlayer> players;
+
+  private long lastTick;
+  private long currentDelay;
+  private long spawnedEntities;
+
+  public Spawner(SpawnerDefinition definition, Match match) {
+    this.definition = definition;
+    this.match = match;
+
+    this.lastTick = match.getTick().tick;
+    this.players = new OnlinePlayerMapAdapter<>(PGM.get());
+    generateDelay();
+  }
+
+  @Override
+  public String getId() {
+    return definition.id;
+  }
+
+  @Override
+  public SpawnerDefinition getDefinition() {
+    return definition;
+  }
+
+  @Override
+  public void tick(Match match, Tick tick) {
+    if (!canSpawn()) return;
+    if (match.getTick().tick - lastTick >= currentDelay) {
+      for (Spawnable spawnable : definition.objects) {
+        final Vector location = definition.spawnRegion.getRandom(match.getRandom());
+        spawnable.spawn(location.toLocation(match.getWorld()), match);
+        match
+            .getWorld()
+            .spigot()
+            .playEffect(
+                location.toLocation(match.getWorld()), Effect.FLAME, 0, 0, 0, 0.15f, 0, 0, 40, 64);
+
+        if (spawnable.isTracked()) {
+          spawnedEntities = spawnedEntities + spawnable.getSpawnCount();
+        }
+      }
+      generateDelay();
+      lastTick = match.getTick().tick;
```

Comment:
I think this should be handled in `generateDelay`. Maybe rename the method `calculateDelay`